### DISCUSSION TITLE
Is the AWS SDK client expected to be able to work with minio?

### DISCUSSION THREAD

**Discussion by io7m** [minio/minio#20298] [Q&A] [answered]
From the command line (don't worry about secrets, these are all just temporary local development credentials):

```
$ mc put /tmp/example.txt local/general/example.txt
/tmp/example.txt:    7 B / 7 B ┃▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓┃ 1.76 KiB/s 0s
```

This results in a `PUT` request that looks like this:

```
PUT /general/example.txt HTTP/1.1
Host: localhost:9000
User-Agent: MinIO (linux; amd64) minio-go/v7.0.73 mcli/RELEASE.2024-07-31T15-58-33Z
Content-Length: 179
Accept-Encoding: zstd,gzip
Authorization: AWS4-HMAC-SHA256 Credential=lQtpoZ5cpbJtwyDtHvub/20240821/us-east-1/s3/aws4_request,SignedHeaders=host;x-amz-content-sha256;x-amz-date;x-amz-decoded-content-length,Signature=343b356c36e078349baec73d7068c5dd68eaca18307540e321c7020848094fe3
Content-Type: text/plain
X-Amz-Content-Sha256: STREAMING-AWS4-HMAC-SHA256-PAYLOAD
X-Amz-Date: 20240821T110822Z
X-Amz-Decoded-Content-Length: 7

7;chunk-signature=cfbaeb5275f184d4f6dad0f4e209b4b921c20e1e542cf06c96acdec3d723ec08
Hello!

0;chunk-signature=155db36a772746b7d733babb84eb8d1f50f6f9f965ae6ead7d864f0d4e1da361

```

Using the AWS SDK client, a `PutObject` request using the same credentials results in this `PUT` request:

```
PUT /example.txt HTTP/1.1
Host: general.localhost:9000
Transfer-encoding: chunked
amz-sdk-invocation-id: 62959d84-7481-dc3a-5561-7395d265a7e6
amz-sdk-request: attempt=1; max=4
Authorization: AWS4-HMAC-SHA256 Credential=lQtpoZ5cpbJtwyDtHvub/20240821/us-east-1/s3/aws4_request, SignedHeaders=amz-sdk-invocation-id;amz-sdk-request;content-encoding;content-length;content-type;host;x-amz-checksum-sha256;x-amz-content-sha256;x-amz-date;x-amz-decoded-content-length, Signature=8aeb31344f54251e12dc7588ebb9cbfd3f9429ad0947068a3440c641b3a8d068
Content-Encoding: aws-chunked
Content-Type: text/plain
User-Agent: aws-sdk-java/2.27.8 Linux/6.9.6-arch1-1 OpenJDK_64-Bit_Server_VM/22.0.2+9 Java/22.0.2 vendor/Arch_Linux io/sync http/UNKNOWN cfg/retry-mode/legacy cfg/auth-source#stat
x-amz-checksum-sha256: b22b009134622b6508d756f1062455d71a7026594eacb0badf81f4f677929ebe
x-amz-content-sha256: STREAMING-AWS4-HMAC-SHA256-PAYLOAD
X-Amz-Date: 20240821T110829Z
x-amz-decoded-content-length: 7

1
7
1
;
f
chunk-signature
1
=
40
0edc71f38c9b4e6491787e44ce6e40f3fbe976f3aead98b1f163c2353cf25f12
2


7
Hello!

2


1
0
1
;
f
chunk-signature
1
=
40
b631f0f30817f0c1dd8928bf66ddde47e99dc7eb8ff00f1c97af8c7348fc193d
2


2


0


```

This results in the following from the server:

```
<?xml version="1.0" encoding="UTF-8"?>
<Error><Code>SignatureDoesNotMatch</Code><Message>The request signature we calculated does not match the signature you provided. Check your key and signing method.</Message><Key>example.txt</Key><BucketName>general</BucketName><Resource>/example.txt</Resource><RequestId>17EDB9CA03FCD960</RequestId><HostId>dd9025bab4ad464b049177c95eb6ebf374d3b3fd1af9251148b658df7ac2e3e8</HostId></Error>
```

There are obviously differences.

1. The AWS SDK uses `Transfer-encoding: chunked`. I don't know if this affects signature calculations.
2. I'm using virtualhost style buckets, but `mc` seems to always use path style buckets. I don't think this should make a difference.
3. I don't know enough about what's going on internally, but the payload in the AWS client looks very different to me.

Is it expected that I should be able to use the AWS S3 SDK?

**Comment by klauspost** [comment]
First use "path style" for AWS SDK. `general.localhost:9000` doesn't make sense and is not supported.

The printout looks to include some unrelated numbers and characters.

`X-Amz-Date` included in the chunked signature, so it will be different unless you fix the values of all headers.

**Reply by io7m** [reply]
> First use "path style" for AWS SDK. general.localhost:9000 doesn't make sense and is not supported.

Interestingly, it does "work" for some requests. If I ask to list objects in a bucket using virtualhost style requests, `minio` does the right thing. What do you mean by "unsupported" exactly? Is this "it might work, but we don't test it and don't intend anyone to use it"?

> The printout looks to include some unrelated numbers and characters.

It's a text dump taken directly from `wireshark`. That's exactly what the request was, as strange as it may seem.

> X-Amz-Date included in the chunked signature, so it will be different unless you fix the values of all headers.

I'm not sure what you mean by this. These aren't headers I set myself, these are just what both clients did when I made a logically equivalent request from each.

**Reply by klauspost** [reply]
>  What do you mean by "unsupported" exactly?

That you shouldn't use it.

I don't know what wireshark is showing you, but it is not valid: https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-streaming.html

> I'm not sure what you mean by this

I mean that signatures will be different for different requests.

**Reply by io7m** [reply]
Switching to path-based fails in the same manner (a signature error).

> I don't know what wireshark is showing you, but it is not valid: https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-streaming.html

Looking at it a bit more closely, it appears to me that it's a chunked encoding embedded inside another chunked encoding, hence the weird appearance. I've attached a zip file containing the exact bytes of the request that the AWS client sent to `minio` over the wire.

I don't know if the AWS SDK or `minio` is at fault here. I'm not sure anything is "invalid", but maybe both sides don't agree on what the spec means.

[request.zip](https://github.com/user-attachments/files/16691300/request.zip)

**Comment by marktheunissen** [comment]
Hi @io7m the answer to your question "Is the AWS SDK client expected to be able to work with minio?" is yes, and we have testing suites the use the AWS SDK to validate it works against MinIO, please see: https://github.com/minio/mint. These are not totally comprehensive but they can be extended easily to cover whichever cases we need.

For any debugging you'll need to provide much more info and specifics, the wireshark dumps are a start but more is needed. What do you mean by "Using the AWS SDK client"? Which language? Please post all the code.

**Comment by io7m** [comment, marked as answer]
That's good to know, thanks.

In my case, it turned out to be a faulty configuration on the client side. I was using an HTTP client implementation based on `java.net.http` rather than the Apache or Netty based clients the AWS SDK ships with. Evidently, `minio` doesn't really like people using `Transfer-Encoding: chunked`; it results in the "chunked encoding inside a chunked encoding" situation shown above.

If the client is forced into _not_ doing this, everything seems to work properly.
